sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -82,8 +109,142 @@ case class GpuIf(
     }
   }
 
-  override def columnarEval(batch: ColumnarBatch): Any = computeIfElse(batch, predicateExpr,
-    trueExpr, falseExpr.columnarEval(batch))
+  override def columnarEval(batch: ColumnarBatch): Any = {
+
+    val gpuTrueExpr = trueExpr.asInstanceOf[GpuExpression]
+    val gpuFalseExpr = falseExpr.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuExpressionsUtils.columnarEvalToColumn(predicateExpr, batch)) { pred =>
+      if (isAllTrue(pred)) {
+        trueExpr.columnarEval(batch)
+      } else if (isAllFalse(pred)) {
+        falseExpr.columnarEval(batch)
+      } else if (gpuTrueExpr.hasSideEffects || gpuFalseExpr.hasSideEffects) {
+        withResource(GpuColumnVector.from(batch)) { tbl =>
```

Comment:
It would be nice to factor this out to a separate method so it's easier to read the `columnarEval` code at a higher level, and it also has the nice side-effect (ha!) of allowing us to give the algorithm a nice name and some comments before it explaining at a high level what's happening.